Store block interval in genesis block

kc1212 opened the pull request:
Store block interval in genesis configuration so that applications, such as eventlog, can set their own block intervals.

ineiti commented on omniledger/service/contracts.go:
```diff
@@ -19,6 +20,9 @@ var ZeroDarc = darc.ID(make([]byte, 32))
 var ConfigID = ObjectID{ZeroDarc, ZeroNonce}
 
+// BlockIntervalID is the key to the block interval information.
+var BlockIntervalID = ObjectID{darc.ID([]byte("block_interval")), ZeroNonce}
```
I would put the block interval into the `ConfigID` object/account. We will have other things that we want to configure, and if we create another object for each configuration, there might be too many objects.

So the `ConfigID` object has a value that is a marshalled `struct{ GenesisDarc darc.Darc, BlockInterval time.Duration}`. And we need to find a good way to describe the data-structures stored in the contracts to access them from the outside.

kc1212 replied:
The reason I created a new ID is for that exact reason. We will need to unmarshal all of the config to just read one int64. I can try to keep it in memory upon creation.

ineiti replied:
OK, I see. I was worried about marshalling/unmarshalling for events that come in at 100s per second. For the configuration, it will be much less, so I think it's more convenient to have a `Configuration` structure that is marshalled/unmarshalled as needed.

When we discussed the ConfigID first, we only thought about the genesis-darc, but now other data will also have to be stored. One thing we could do is the following (plus renaming the variables ;):

## 1 - One Object for all configuration

`var GenesisDarcIDObjectID = ObjectID{ZeroDarc, ZeroNonce}` holds the ID of the genesis-darc
`var GenesisDarcObjectID = ObjectID{GenesisDarcID, ZeroNonce}` holds the genesis-darc itself so that it can be evolved
`var ConfigObjectID = ObjectID{GenesisDarcID, OneNonce}` holds the configuration (BlockInterval, Roster(?), IdentitySkipchain, CoinsMapping (CPU, Memory), ...)

So if you want to look up the configuration, you have to go through the `GenesisDarcIDObjectID` (bad name, but it explains nicely what it is), to get the `ConfigObjectID` and then you can read the configuration.

## 2 - One object for every configuration parameter

The alternative you propose is to have
`var ConfigIntervalObjectID = ObjectID{GenesisDarc, OneNonce}`
`var ConfigRosterObjectID = ObjectID{GenesisDarc, TwoNonce}`
...
and then each `Config.*ObjectID` will have to have its own contract that defines how you're allowed to change that object.

## Conclusion

I prefer 1) because it needs less contracts and gives a 'central' configuration that you can change. Using the actions in the genesis-darc you can define who is allowed to change which configuration parameter, as each parameter will have its own command (`UpdateInterval`, `UpdateRoster`, ...)

ineiti commented on omniledger/service/contracts.go:
```diff
@@ -41,9 +59,32 @@ func (s *Service) ContractConfig(cdb collection.Collection, tx Instruction, coin
 		log.Error("couldn't verify darc")
 		return
 	}
+
```
Later on we will need to add more than `Spawn` to the Config-contract.

ineiti commented on omniledger/service/service.go:
```diff
@@ -73,8 +75,9 @@ type Service struct {
-// TODO: this should go into the genesis-configuration
-var waitQueueing = 5 * time.Second
+// defaultInterval is used if the BlockInterval field in the genesis
+// trnasaction is not set.
```
s/trnas/trans/

ineiti commented on omniledger/service/service.go:
```diff
@@ -421,13 +417,69 @@ func (s *Service) db() *skipchain.SkipBlockDB {
+	if string(contract) != ContractConfigID {
+		return nil, errors.New("did not get " + ContractConfigID)
+	}
+	if len(val) != 64 {
```
Why is this not 32 bytes? If you put the darc-id in the val, then it's only 32 bytes. And the darc is by definition stored in the `[darcID:zeroNonce]`.

ineiti commented on omniledger/service/service.go:
```diff
@@ -421,13 +417,69 @@ func (s *Service) db() *skipchain.SkipBlockDB {
+	if colldb == nil {
+		return nil, fmt.Errorf("collection for skipchain ID %s does not exist", sid.Short())
+	}
+	value, contract, err := colldb.GetValueContract(padKey(dID))
```
We need to find a better name for `padKey` - I wrote it when testing. Shall we call it `getDarcKey`?

kc1212 commented on omniledger/service/contracts.go:
```diff
@@ -41,9 +59,32 @@ func (s *Service) ContractConfig(cdb collection.Collection, tx Instruction, coin
+
 	return []StateChange{
-		NewStateChange(Create, ConfigID, ContractConfigID, tx.ObjectID.DarcID),
+		NewStateChange(Create, GenesisReferenceID, ContractConfigID, tx.ObjectID.DarcID),
```
I don't understand this part, we don't have a `ContractReferenceID`.

kc1212 commented:
@ineiti Sorry that was sloppy of me. I thought if tests all pass then the code is most likely to be correct, but didn't realise the `LoadConfig` function wasn't getting tested. I've added a test and the problems that you pointed out were evident. Hopefully my latest commit addresses those issues.
